truncated_MonteCarlo.py

```diff
@@ -0,0 +1,96 @@
+"""
+Created on Fri Nov  29 18:04:13 2019
+A script to  run simulations with the truncated Monte-Carlo method 
+
+@author: Thomas-Galtier
+"""
+ 
+import numpy as np 
+import scenario
+
+from scipy.stats import norm
+
+
+def truncated_MC(preprocessed_node_list, sv_accuracy=0.001, contrib_accuracy=0.01):
+    """Return the vector of approximated shapeley value corresponding to a list of node and a characteristic function using the truncated monte-carlo method."""
+    
+    characteristic_func=fl_training.compute_test_score
+    n = len(preprocessed_node_list)
+    
+    # we store the value of the characteristic function in order to avoid recomputing it twice
+    char_value_dict={():0} # the dictionary that will countain the values
+    listn=np.arange(n)
+    
+    # return the characteristic function of the nodelist associated to the ensemble  permut, without recomputing it if it was already computed
+    def not_twice_characteristic(permut):
+        #sort permut
+        isin=np.repeat(False,n)
+        isin[permut]=True
+        permut=listn[isin]
+        try: #return the characteristic_func(permut) if it was already computed
+            return char_value_dict[tuple(permut)]
+        except KeyError: #characteristic_func(permut) has not been computedyet, so we compute, store, and return characteristic_func(permut)
+            small_node_list = [preprocessed_node_list[i] for i in permut]
+            char_value_dict[tuple(permut)]= characteristic_func(small_node_list)
+            return char_value_dict[tuple(permut)]
+
+        
+    # truncated_MC initialisation
+    if n==1:
+        return np.array([1])
+    else:
+        characteristic_all_node= not_twice_characteristic(np.arange(n)) #characteristic function on all nodes
+        sv = np.zeros(n) #store current estimation of the shapley values 
+        mean_squares_sv = np.zeros(n) # used for variance computation in the condition of the while loop below 
+        permutation=np.zeros(n) # store the current permutation 
+        firsts_are_equal =True
+        t=0
+        while t<50 or t<=norm.ppf(0.975, loc=0, scale=1)**2* np.max(mean_squares_sv-sv**2) /sv_accuracy**2 or firsts_are_equal: #check if the length of the confidence interval  is below the value of sv_accuracy
```

Comment:
Shouldn't it be an `and` condition?

Reply:
In `t<=norm.ppf(0.975, loc=0, scale=1)**2* np.max(mean_squares_sv-sv**2) /sv_accuracy**2` I don't understand what this quantity represents and why/how you relate it to `t`. Could you please explain?

Reply:
No it is well an `or` condition. For two reasons: 
1) The approximations of the Shapeley values relies on the Central limit theorem which is an asymptotic result. We want to have at least `50` iterations so that we can say we are in an asymptotic situation, and that our approximations are located around the true Shapeley values.
2) The quantity `mean_squares_sv-sv**2` is supposed to asymptotically approximate the variance of the contributions so `t` needs to be at least 50 so that this approximation (and therefore the second condition) is meaningfull.

Reply:
For the second condition I think we should change it because it it too strong...  Nevertheless the idea behind this condition is to make sure that half the length of of the confidence interval for each approximation of a Shapeley value is lower than `sv_accuarcy`, so that the approximation errors are below `sv_accuarcy ` with 95% chance.

let`q=norm.ppf(0.975, loc=0, scale=1)` and `v=mean_squares_sv-sv**2`. Let `L[i]` be half the length of the confidence interval for the i-th Shapeley value approximation then `L[i]=q*np.sqrt(v[i]/t)`. The condition `t<=norm.ppf(0.975, loc=0, scale=1)**2* np.max(mean_squares_sv-sv**2) /sv_accuracy**2` is equivalent to `np.max(L)<sv_accuracy`.

Reply:
This condition impose a very high number of iterations so I think we should change it. We should probably replace tit by something like `np.max(np.mean(contributions,axis=0) - np.mean(contributions[0:t-1],axis=0))>sv_accuracy`

Reply:
I finally kept the original condition.